content/blog/container-filesystems/index.md

```diff
@@ -0,0 +1,132 @@
+    "UpperDir": "/var/lib/docker/overlay2/63ec1a08b063c0226141a9071b5df7958880aae6be5dc9870a279a13ff7134ab/diff",
+    "WorkDir": "/var/lib/docker/overlay2/63ec1a08b063c0226141a9071b5df7958880aae6be5dc9870a279a13ff7134ab/work"
+  },
+  "Name": "overlay2"
+}
+```
+
+Let’s break this down:
+ - `LowerDir`: Includes the filesystems of all the layers inside the container except the last one
+ - `UpperDir`: The filesystem of the top-most layer of the container. This is also where any run-time modifications are reflected.
+ - `MergedDir`: A combined view of all the layers of the filesystem.
+ - `WorkDir`: An internal working directory used to manage the filesystem.
+
+::: div image-xl
+<svg title='Structure of container filesystems based on overlayfs.' src='overlayfs.png' />
+:::
+
+So to see the files inside our container, we simply need to look at the MergedDir path.
+
+```
+sudo ls /var/lib/docker/overlay2/63ec1a08b063c0226141a9071b5df7958880aae6be5dc9870a279a13ff7134ab/merged
+```
+
+If you want to learn in more detail how the filesystem works, you can check out this excellent blog post on the overlay filesystem by Martin Heinz: [https://martinheinz.dev/blog/44](https://martinheinz.dev/blog/44)
+
+## Method 4: /proc/&lt;pid&gt;/root
+
+Saving the best for last, there’s an even easier way to find the container’s filesystem from the host. Using the host PID of a process inside the container, you can simply run:
+
+```
+sudo ls /proc/<pid>/root
+```
+
+Linux has taken care of giving you a view into the mount namespace of the process.
+
+At this point, you’re probably thinking: why didn’t we just lead with this approach and made it a one-line blog post...but it’s all about the journey, right?
+
+## Bonus: /proc/&lt;pid&gt;/mountinfo
+For the curious, all the information about the container’s overlay filesystem discussed in Method 3 can also be discovered directly from the Linux `/proc` filesystem. If you simply look at `/proc/<pid>/mountinfo`, you’ll see something like this:
+
+```
+2363 1470 0:90 / / rw,relatime master:91 - overlay overlay rw,lowerdir=/var/lib/docker/overlay2/l/YZVAVZS6HYQHLGEPJHZSWTJ4ZU:/var/lib/docker/overlay2/l/ZYW5O24UWWKAUH6UW7K2DGV3PB,upperdir=/var/lib/docker/overlay2/63ec1a08b063c0226141a9071b5df7958880aae6be5dc9870a279a13ff7134ab/diff,workdir=/var/lib/docker/overlay2/63ec1a08b063c0226141a9071b5df7958880aae6be5dc9870a279a13ff7134ab/work
+2364 2363 0:93 / /proc rw,nosuid,nodev,noexec,relatime - proc proc rw
+2365 2363 0:94 / /dev rw,nosuid - tmpfs tmpfs rw,size=65536k,mode=755,inode64
+…
+```
+
+Here you can see that the container has mounted an overlay filesystem as its root. It also reports the same type of information that `docker inspect` reports, including the LowerDir and UpperDir of the container’s filesystem. It’s not directly showing the MergedDir, but you can just take the UpperDir and change `diff` to `merged`, and you have a view into the filesystem of the container.
+
+## How we use this at Pixie
```

Comment:
Should this paragraph be put closer to the top? This can be a good introduction on how the problem was discovered and motivated.

Reply:
Do you mean at the top of the entire blog or the top of the section?

Top of the blog would mean it goes too deep too quickly, and the blog is not just about overlay filesystems.